bpf: extract code duplication from identity helper functions

viktor-kurchenko opened the pull request:
Please see description per commit.

P.S.
This is actually an attempt to address this comment: https://github.com/cilium/cilium/pull/43258#pullrequestreview-3566621790

viktor-kurchenko committed:
bpf, identity: simplify set_identity_mark function

It looks like the following instruction is redundant and can be safely
removed: `ctx->mark &= MARK_MAGIC_KEY_MASK`.
The reason is that we apply the same logic in the line above:
`ctx->mark = (magic & MARK_MAGIC_KEY_MASK)`.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

viktor-kurchenko committed:
bpf, clustermesh: minor code refactoring

Remove unneeded variable in the `get_identity_len` function and redundant
brackets in the `ctx_set_cluster_id_mark` function.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

viktor-kurchenko committed:
bpf, identity: extract set cluster id logic

The commit extracts duplicated code from the `set_identity_mark` and
`ctx_set_cluster_id_mark` functions in the new `ctx_set_cluster_id`
helper function.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

viktor-kurchenko commented:
/test

julianwiedmann commented:
I'd just want to make sure that we have a common understanding of what the code even does :)
1. `set_identity_mark()` formats the `skb->mark` with `identity`, `cluster_id` and one of the `magic` identifiers
2. `set_cluster_id_mark()` formats the skb->mark with **only** the `cluster_id` and the `MARK_MAGIC_CLUSTER_ID` identifier.

Is that correct? Is it accurate to say that `set_cluster_id_mark()` is merely a special case of `set_identity_mark()`, with the `identity` part set to 0?

julianwiedmann commented:
> > So if both variants currently essentially share the same bit-layout, I wonder if we should codify this and have a shared helper - `encode_identity_and_cluster_id(identity, cluster_id, magic)` ?
> 
> Sounds reasonable to me. One possibly confusing bit of that approach is that an identity actually encodes both the cluster_id and the cluster-local identity (i.e., the lower 16 bits, or 15 with extended clustermesh), so we need to be careful with naming.

`encode_local_identity_and_cluster_id(local_identity, cluster_id, magic)` then maybe? If we then end up having multiple callers that need to perform the identity/cluster_id split (in order to provide the parameters in the expected format), we can always add a helper for that as well.

viktor-kurchenko commented:
/test

viktor-kurchenko committed:
bpf, identity: format_cluster_id_mark instead of ctx_set_cluster_id

The commit renames `ctx_set_cluster_id` function into
`format_cluster_id_mark`. The function now returns cluster ID lower and
upper bits instead of manipulating them in context directly.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>
